lib/puppet-lint/plugins/check_classes.rb

```diff
@@ -100,42 +99,46 @@ def check
 # https://docs.puppet.com/guides/style_guide.html#display-order-of-parameters
 PuppetLint.new_check(:parameter_order) do
   def check
-    (class_indexes + defined_type_indexes).each do |class_idx|
-      unless class_idx[:param_tokens].nil?
-        paren_stack = []
-        hash_or_array_stack = []
-        class_idx[:param_tokens].each_with_index do |token, i|
-          if token.type == :LPAREN
-            paren_stack.push(true)
-          elsif token.type == :RPAREN
-            paren_stack.pop
-          elsif token.type == :LBRACE || token.type == :LBRACK
-            hash_or_array_stack.push(true)
-          elsif token.type == :RBRACE || token.type == :RBRACK
-            hash_or_array_stack.pop
-          end
-          next if (! hash_or_array_stack.empty?)
-          next unless paren_stack.empty?
+    (class_indexes + defined_type_indexes).reject { |class_idx|
+      class_idx[:param_tokens].nil?
+    }.each do |class_idx|
+      paren_stack = []
+      hash_or_array_stack = []
+      class_idx[:param_tokens].each_with_index do |token, i|
+        if token.type == :LPAREN
+          paren_stack.push(true)
+        elsif token.type == :RPAREN
+          paren_stack.pop
+        elsif token.type == :LBRACE || token.type == :LBRACK
+          hash_or_array_stack.push(true)
+        elsif token.type == :RBRACE || token.type == :RBRACK
+          hash_or_array_stack.pop
+        end
 
-          if token.type == :VARIABLE
-            if token.next_code_token.nil? || [:COMMA, :RPAREN].include?(token.next_code_token.type)
-              prev_tokens = class_idx[:param_tokens][0..i]
-              unless prev_tokens.rindex { |r| r.type == :EQUALS }.nil?
-                unless token.prev_code_token.nil? or token.prev_code_token.type == :EQUALS
-                  msg = 'optional parameter listed before required parameter'
-                  notify :warning, {
-                    :message => msg,
-                    :line    => token.line,
-                    :column  => token.column,
-                  }
-                end
-              end
+        next unless hash_or_array_stack.empty? && paren_stack.empty?
+
+        if required_parameter?(token)
+            prev_tokens = class_idx[:param_tokens][0..i]
+            unless prev_tokens.rindex { |r| r.type == :EQUALS }.nil?
+              msg = 'optional parameter listed before required parameter'
+              notify :warning, {
+                :message => msg,
+                :line    => token.line,
+                :column  => token.column,
+              }
             end
-          end
         end
       end
     end
   end
+
+  def required_parameter?(token)
+    return unless token.type == :VARIABLE
```

Comment:
The general convention for Ruby predicate methods is to return a boolean, so I'd suggest changing this line to `return false unless token.type == :VARIABLE` so that it'll return `false` in this case instead of `nil`.